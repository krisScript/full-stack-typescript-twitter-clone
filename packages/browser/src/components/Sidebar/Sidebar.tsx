import React, { FC, useState } from 'react';
import { observer } from 'mobx-react-lite';
import { FontAwesomeIcon } from '@fortawesome/react-fontawesome';
import {
  StyledSidebar,
  Backdrop,
  Container,
  AuthenticationBar,
  AuthenticatedSidebarHeader,
  SidebarHeader,
  LogoContainer,
  SidebarBody,
  SidebarList,
  SidebarButton,
} from './StyledSidebar';
import Logo from 'components/Logo/Logo';
import SearchBar from 'components/SearchBar/SearchBar';
import AuthState from 'types/AuthState';
import Avatar from 'styled/Avatar';

interface SidebarProps {
  isActive: boolean;
  authState: AuthState;
  theme: 'light' | 'dark';
  resetAuthState: () => void;
  toggleTheme: () => void;
  toggleSidebar: () => void;
}

export const Sidebar: FC<SidebarProps> = ({
  isActive,
  authState,
  theme,
  resetAuthState,
  toggleTheme,
  toggleSidebar,
}) => {
  const { user, isAuth, token } = authState;
  const { username, handle, following, followers } = user;
  console.log(following, followers);
  return (
    <StyledSidebar on={isActive}>
      <Container>
        {authState.isAuth ? (
          <AuthenticatedSidebarHeader>
            <Avatar size="large">
              <img src="data:image/svg+xml;base64,PHN2ZyB4bWxucz0iaHR0cDovL3d3dy53My5vcmcvMjAwMC9zdmciIHZpZXdCb3g9Ii0xMS41IC0xMC4yMzE3NCAyMyAyMC40NjM0OCI+CiAgPHRpdGxlPlJlYWN0IExvZ288L3RpdGxlPgogIDxjaXJjbGUgY3g9IjAiIGN5PSIwIiByPSIyLjA1IiBmaWxsPSIjNjFkYWZiIi8+CiAgPGcgc3Ryb2tlPSIjNjFkYWZiIiBzdHJva2Utd2lkdGg9IjEiIGZpbGw9Im5vbmUiPgogICAgPGVsbGlwc2Ugcng9IjExIiByeT0iNC4yIi8+CiAgICA8ZWxsaXBzZSByeD0iMTEiIHJ5PSI0LjIiIHRyYW5zZm9ybT0icm90YXRlKDYwKSIvPgogICAgPGVsbGlwc2Ugcng9IjExIiByeT0iNC4yIiB0cmFuc2Zvcm09InJvdGF0ZSgxMjApIi8+CiAgPC9nPgo8L3N2Zz4K" />
            </Avatar>
            <h3>{username}</h3>
            <h4>@{handle}</h4>
            <div>
              <button>
                <span>0</span> Followers
              </button>
              <button>
                <span>{following.length}</span> Following
              </button>
            </div>
          </AuthenticatedSidebarHeader>
        ) : (
          <SidebarHeader>
            <LogoContainer>
              <Logo type="vertical" />
            </LogoContainer>
            <AuthenticationBar>
              <button>Log In</button>
              <span>or</span>
              <button>Sign Up</button>
            </AuthenticationBar>
          </SidebarHeader>
        )}
        <SidebarBody>
          <SearchBar />
          <SidebarList>
            <li>
              <SidebarButton>
                <span>
                  <FontAwesomeIcon icon={'home'} />
                </span>
                <p>Home</p>
              </SidebarButton>
            </li>
            {authState.isAuth ? (
              <>
                {' '}
                <li>
                  <SidebarButton>
                    <span>
                      <FontAwesomeIcon icon={'user'} />
                    </span>
                    <p>Profile</p>
                  </SidebarButton>
                </li>
                <li>
                  <SidebarButton>
                    <span>
                      <FontAwesomeIcon icon={'bookmark'} />
                    </span>
                    <p>Bookmarks</p>
                  </SidebarButton>
                </li>
              </>
            ) : (
              ''
            )}
            <li>
              <SidebarButton
                onClick={toggleTheme}
                data-testid={'theme-button'}
              >
                {theme === 'light' ? (
                  <>
                    <span>
                      <FontAwesomeIcon icon={'moon'} />
                    </span>
                    <p>Dark Theme</p>
                  </>
                ) : (
                  <>
                    <span>
                      <FontAwesomeIcon icon={'sun'} />
                    </span>
                    <p>Light Theme</p>
                  </>
                )}
              </SidebarButton>
            </li>
          </SidebarList>
          {isAuth ? (
            <SidebarButton onClick={ resetAuthState}>
              <span>
                <FontAwesomeIcon icon={'sign-out-alt'} />
              </span>
              <p>Log Out</p>
            </SidebarButton>
          ) : (
            ''
          )}
        </SidebarBody>
      </Container>
      <Backdrop onClick={toggleSidebar} data-testid="backdrop" />
    </StyledSidebar>
  );
};
export default observer(Sidebar);
