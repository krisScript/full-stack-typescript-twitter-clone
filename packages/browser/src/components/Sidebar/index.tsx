import React, { FC } from 'react';
import { observer } from 'mobx-react-lite';
import { FontAwesomeIcon } from '@fortawesome/react-fontawesome';
import { Link } from 'react-router-dom';
import Logo from 'components/Logo';
import SearchBar from 'components/SearchBar';
import AuthState from 'types/AuthState';
import Avatar from 'components/Avatar';
import UserCredentials from 'components/UserCredentials';
import {
  SidebarWrapper,
  Backdrop,
  Container,
  AuthenticationBar,
  AuthenticatedSidebarHeader,
  SidebarHeader,
  LogoContainer,
  SidebarBody,
  SidebarList,
  SidebarButton,
  SearchBarWrapper,
  AuthenticationButton,
} from './styled';

interface SidebarProps {
  isActive: boolean;
  authState: AuthState;
  theme: 'light' | 'dark';
  resetAuthState: () => void;
  toggleTheme: () => void;
  toggleSidebar: () => void;
}

export const Sidebar: FC<SidebarProps> = ({
  isActive,
  authState: { user },
  theme,
  resetAuthState,
  toggleTheme,
  toggleSidebar,
}) => {
  return (
    <>
      <SidebarWrapper isActive={isActive}>
        <Container>
          {user ? (
            <AuthenticatedSidebarHeader>
              <Link
                to={`/users/${user._id}`}
                data-testid="profile-link-sidebar"
              >
                <Avatar size="medium" avatar={user.avatar} />
              </Link>
              <UserCredentials
                username={user.username}
                handle={user.handle}
                _id={user._id}
                followers={user.followers}
                following={user.following.length}
              />
            </AuthenticatedSidebarHeader>
          ) : (
            <SidebarHeader>
              <LogoContainer>
                <Logo type="vertical" />
              </LogoContainer>
              <AuthenticationBar>
                <AuthenticationButton type="button" onClick={toggleSidebar}>
                  <Link to="/log-in">Log In</Link>
                </AuthenticationButton>
                <span>or</span>
                <AuthenticationButton type="button" onClick={toggleSidebar}>
                  {' '}
                  <Link to="/sign-up">Sign Up</Link>
                </AuthenticationButton>
              </AuthenticationBar>
            </SidebarHeader>
          )}
          <SidebarBody>
            <SearchBarWrapper>
              <SearchBar />
            </SearchBarWrapper>
            <SidebarList>
              <li>
                <Link to="/" onClick={toggleSidebar}>
                  <SidebarButton>
                    <span>
                      <FontAwesomeIcon size="lg" icon="home" />
                    </span>
                    <p>Home</p>
                  </SidebarButton>
                </Link>
              </li>
              {user ? (
                <li>
                  <Link to={`/users/${user._id}`} onClick={toggleSidebar}>
                    <SidebarButton>
                      <span>
                        <FontAwesomeIcon size="lg" icon="user" />
                      </span>
                      <p>Profile</p>
                    </SidebarButton>
                  </Link>
                </li>
              ) : (
                ''
              )}
              <li>
                <SidebarButton onClick={toggleTheme} data-testid="theme-button">
                  <span>
                    <FontAwesomeIcon
                      icon={theme === 'light' ? 'sun' : 'moon'}
                    />
                  </span>
                  <p>{theme} mode</p>
                </SidebarButton>
              </li>
            </SidebarList>
            {user ? (
              <SidebarButton onClick={resetAuthState}>
                <span>
                  <FontAwesomeIcon size="lg" icon="sign-out-alt" />
                </span>
                <p>Log Out</p>
              </SidebarButton>
            ) : (
              ''
            )}
          </SidebarBody>
        </Container>
        <Backdrop onClick={toggleSidebar} data-testid="backdrop" />
      </SidebarWrapper>
    </>
  );
};
export default observer(Sidebar);
