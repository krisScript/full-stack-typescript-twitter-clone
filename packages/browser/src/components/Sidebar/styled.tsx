import styled, { css } from 'styled-components';

interface SidebarProps {
  isActive: boolean;
}
const header = css`
  width: 100%;
  height: 20%;
  min-height: 15rem;
`;
export const SidebarWrapper = styled('aside')<SidebarProps>`
  display: grid;
  grid-template-columns: 3fr 1fr;
  position: fixed;
  z-index: 3;
  width: 100vw;
  height: 100vh;
  ${({ theme }) => theme.mixins.slide};
  ${({ isActive }) =>
    isActive ? `animation: slide 200ms ease-in forwards` : ''};
  @media (max-height: 30rem) {
    overflow: scroll;
  }
  @media ${({ theme }) => theme.mediaQueries.tablet} {
    grid-template-columns: 2fr 3fr;
  }
  @media ${({ theme }) => theme.mediaQueries.desktop} {
    display: none;
  }
`;

export const Backdrop = styled('div')`
  background: ${({ theme }) => theme.transparentBackground};
`;

export const Container = styled('div')`
  background: ${({ theme }) => theme.background};
`;

export const SidebarHeader = styled('div')`
  ${({ theme }) => theme.mixins.header};
  display: grid;
  grid:
    ' . . . ' 1fr
    'logo logo logo ' 3fr
    ' authentication-bar . .' 1fr/1fr 1fr 1fr;
  ${({ theme }) => theme.mixins.border}
`;

export const AuthenticatedSidebarHeader = styled('div')`
  ${({ theme }) => theme.mixins.header}
  display: grid;
  grid-template-columns: 1fr;
  grid-gap: 0.2rem;
  padding-top: 3rem;
  padding-left: 0.7rem;
  ${({ theme }) => theme.mixins.border}
`;

export const SearchBarWrapper = styled('div')`
  margin-left: 0.7rem;
`;
export const LogoContainer = styled('div')`
  display: grid;
  align-content: center;
  justify-items: center;
  grid-area: logo;
`;

export const AuthenticationBar = styled('div')`
  grid-area: authentication-bar;
  display: grid;
  grid-gap: 0.7rem;
  grid-template-columns: repeat(3, auto);
  align-items: center;
  padding-left: 0.5rem;
  width: 15rem;
  span {
    color: ${({ theme }) => theme.primary};
    font-size: 1.4rem;
  }
`;

export const SidebarBody = styled('div')`
  padding-top: 2rem;
  width: 100%;
  height: 80%;
  min-height: 10rem;
`;

export const SidebarList = styled('ul')`
  margin-top: 1.5rem;
  display: flex;
  flex-direction: column;
  width: 100%;
  height: 60%;
  min-height: 30rem;
  margin-bottom: 1rem;
  ${({ theme }) => theme.mixins.border};
  border-top: none;
  border-left: none;
  li {
    width: 10rem;
    margin-top: 1.5rem;
    a {
      display: block;
    }
  }
`;

export const AuthenticationButton = styled('button')`
  ${({ theme }) => theme.mixins.button}
  background:none;
  a {
    font-size: 1.5rem;
    color: ${({ theme }) => theme.color};
  }
`;
export const SidebarButton = styled('button')`
  margin-left: 0.7rem;
  width:auto;
  display: flex;
  flex-flow: row;
  background:none;
  ${({ theme }) => theme.mixins.flexCenter}
  border: none;
  ${({ theme }) => theme.mixins.button}
  color:   ${({ theme }) => theme.color};
   
 
  p {
    display: block;
    font-size: 1.3rem;
  }
  span {
    margin-right: 0.5rem;
    display: block;
    font-size: 1.2rem;
  }
`;
