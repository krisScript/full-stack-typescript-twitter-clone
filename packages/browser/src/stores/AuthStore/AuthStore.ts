import { observable, action } from 'mobx';
import { persist } from 'mobx-persist';
import AuthState from 'types/AuthState';

class AuthStore {
  @persist('object')
  @observable
  public authState: AuthState | null = null;
  @action public setAuthState(newAuthState: AuthState): void {
    this.authState = newAuthState;
  }
  @action public resetAuthState(): void {
    this.authState = null;
  }
}
export default AuthStore;
