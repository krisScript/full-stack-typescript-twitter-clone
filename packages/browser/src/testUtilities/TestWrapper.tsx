import React, { FC } from 'react';
import { BrowserRouter } from 'react-router-dom';
import Theme from 'components/Theme/Theme';

interface TestWrapperProps {
  theme?: 'light' | 'dark';
}
const TestWrapper: FC<TestWrapperProps> = ({ children, theme = 'light' }) => (
  <Theme theme={theme}>
    <BrowserRouter>{children}</BrowserRouter>
  </Theme>
);

export default TestWrapper;
