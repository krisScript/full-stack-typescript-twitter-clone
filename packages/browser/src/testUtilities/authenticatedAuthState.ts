const authenticatedAuthState = {
  isAuth: true,
  user: {
    username: 'mockUserName',
    handle: 'mockUserHandle',
    email: 'mockUserEmail@mail.com',
    profilePhoto: 'default',
    headerPhoto: 'default',
    date: 'mockDate',
    website: 'mockWebsite',
    followers: 0,
    following: [],
    likes: [],
    bookmarks: [],
  },
  token: 'mockToken',
};

export default authenticatedAuthState;
