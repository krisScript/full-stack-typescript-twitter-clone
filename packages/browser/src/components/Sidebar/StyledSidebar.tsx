import styled from 'styled-components';
import { transparentize, darken, lighten } from 'polished';

interface SidebarProps {
  on: boolean;
}
export const StyledSidebar = styled('aside')<SidebarProps>`
  display:grid;
  grid-template-columns:3fr 1fr;
  ${props => (props.theme.testing ? '' : 'position: fixed;')};
  transform: translateX(-100vw);
  z-index: 15;
  top: 0;
  width: 100vw;
  height: 100vh;
  background: none;
  ${props => props.theme.mixins.slide};
  ${props => (props.on ? `animation: slide 200ms linear forwards` : '')}
  @media ${props => props.theme.mediaQueries.tablet} {
    display:none;
  }
`;

export const Backdrop = styled('div')`
  background: ${props => transparentize(0.5, props.theme.secondary)};
`;

export const Container = styled('div')`
  height: 100%;
  background: ${props => props.theme.background};
`;

export const SidebarHeader = styled('div')`
  width: 100%;
  height: 20%;
  display: grid;
  grid:
    ' . . . ' 1fr
    'logo logo logo ' 3fr
    ' authentication-bar . .' 1fr/1fr 1fr 1fr;
  ${props => props.theme.mixins.border}
`;

export const AuthenticatedSidebarHeader = styled('div')`
  width: 100%;
  height: 20%;
  display: grid;
  grid-template-columns: 1fr;
  grid-gap: 0.2rem;
  padding-top: 3rem;
  padding-left: 0.5rem;
  ${props => props.theme.mixins.border}
  
  h3 {
    font-size: 1.3rem;
    color: ${props =>
      props.theme.currentTheme === 'light'
        ? props.theme.color
        : props.theme.white};
  }
  h4 {
    font-size: 1.1rem;
    color: ${props => props.theme.secondary};
  }
  button {
    font-size: 1.1rem;
    font-weight: bold;
    background: none;
    ${props => props.theme.mixins.button}
    color: ${props => props.theme.secondary};
    span {
      color: ${props =>
        props.theme.currentTheme === 'light'
          ? darken(0.4, props.theme.secondary)
          : lighten(0.2, props.theme.secondary)};
    }
    margin-right: 0.5rem;
  }
`;

export const LogoContainer = styled('div')`
  display: grid;
  align-content: center;
  justify-items: center;
  grid-area: logo;
`;
export const AuthenticationBar = styled('div')`
  grid-area: authentication-bar;
  display: grid;
  grid-gap: 0.3rem;
  grid-template-columns: repeat(3, auto);
  align-items: center;
  padding-left: 0.5rem;
  max-width: 15rem;
  * {
    font-size: 1.5rem;
    font-weight: bold;
  }
  span {
    display: block;
    text-align: center;
    color: ${props => props.theme.primary};
  }
  button {
    ${props => props.theme.mixins.button}
    background: none;
    color: ${props =>
      props.theme.currentTheme === 'light'
        ? props.theme.color
        : props.theme.white};
  }
`;

export const SidebarBody = styled('div')`
  padding-top: 2rem;
  width: 100%;
  height: 80%;
`;

export const SidebarList = styled('ul')`
  margin-top: 1.5rem;
  display: flex;
  flex-direction: column;
  width: 100%;
  height: 60%;
  margin-bottom: 1rem;
  ${props => props.theme.mixins.border};
  border-top: none;
  border-left: none;
  li {
    width: 10rem;
    margin-top: 1.5rem;
  }
`;

export const SidebarButton = styled('button')`
  margin-left: 0.5rem;
  width:auto;
  display: flex;
  flex-flow: row;
  background:none;
  border: none;
  ${props => props.theme.mixins.button}
  color: ${props =>
    props.theme.currentTheme === 'light'
      ? props.theme.color
      : props.theme.white};
   
  p {
    display: block;
    font-size: 1.3rem;
  }
  span {
    margin-right: 0.5rem;
    display: block;
    font-size: 1.2rem;
  }
`;
