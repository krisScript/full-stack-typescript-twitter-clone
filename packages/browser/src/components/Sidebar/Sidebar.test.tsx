import React from 'react';
import { render, waitForElement, getByTestId } from '@testing-library/react';
import '@testing-library/jest-dom/extend-expect';
import UserEvent from '@testing-library/user-event';
import Sidebar from './Sidebar';
import { defaultAuthState } from 'stores/AuthStore/AuthStore';
import userEvent from '@testing-library/user-event';
import TestWrapper from 'testUtilities/TestWrapper';
import authenticatedAuthState from 'testUtilities/authenticatedAuthState';

describe('Sidebar', () => {
  const theme = 'light';
  const toggleTheme = jest.fn();
  const resetAuthState = jest.fn();
  const toggleSidebar = jest.fn();
  it('render Sidebar', async () => {
    expect.assertions(11);

    const { container, getByText, rerender } = render(
      <Sidebar
        toggleSidebar={toggleSidebar}
        isActive={true}
        authState={defaultAuthState}
        resetAuthState={resetAuthState}
        theme={theme}
        toggleTheme={toggleTheme}
      />,
      {
        wrapper: ({ children }) => <TestWrapper children={children} />,
      },
    );
    const themeButton = await waitForElement(() =>
      getByTestId(container, 'theme-button'),
    );
    const logInButton = await waitForElement(() => getByText('Log In'));
    const signUpButton = await waitForElement(() => getByText('Sign Up'));
    const backdrop = await waitForElement(() =>
      getByTestId(container, 'backdrop'),
    );

    userEvent.click(backdrop);
    UserEvent.click(themeButton);

    expect(container).toBeTruthy();
    expect(themeButton).toBeTruthy();
    expect(themeButton.textContent).toMatchSnapshot();
    expect(logInButton).toBeTruthy();
    expect(signUpButton).toBeTruthy();
    expect(toggleTheme).toHaveBeenCalledTimes(1);
    expect(toggleSidebar).toHaveBeenCalledTimes(1);

    rerender(
      <Sidebar
        toggleSidebar={toggleSidebar}
        isActive={true}
        resetAuthState={resetAuthState}
        authState={authenticatedAuthState}
        theme={'light'}
        toggleTheme={toggleTheme}
      />,
    );
    const lightThemeButton = await waitForElement(() =>
      getByTestId(container, 'theme-button'),
    );
    const logOutButton = await waitForElement(() => getByText('Log Out'));

    UserEvent.click(logOutButton);

    expect(themeButton.textContent).toMatchSnapshot();
    expect(lightThemeButton).toBeTruthy();
    expect(logOutButton).toBeTruthy();
    expect(resetAuthState).toHaveBeenCalledTimes(1);
  });
});
