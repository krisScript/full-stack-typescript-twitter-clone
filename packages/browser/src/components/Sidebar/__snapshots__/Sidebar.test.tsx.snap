// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Sidebar render Sidebar 1`] = `"Dark Theme"`;

exports[`Sidebar render Sidebar 2`] = `"Dark Theme"`;
